### DISCUSSION TITLE
Java 17 with keeper 0.14.0

### DISCUSSION THREAD

**Discussion by igor-budi** [slackhq/keeper#121] [Q&A] [answered]
I am trying to migrate to the new AGP 8 so I updated keeper to 0.14.0 as well.
My project uses toolchain config set to 17 and all the target and source for java and kotlin are also set to 17.

With this setup, I can't complete a build
```
One or more classes have class file version >= 56 which is not offi****ally supported.                                                                                                                                            
Error in /codefresh/volume/Android/*****/app/build/intermediates/keeper/************AndroidTest/classes.jar:[REDUCTED].class:                                              
com.android.tools.r8.internal.nc: Sealed classes are not supported as program classes  
```

I did find this SO
https://stackoverflow.com/questions/73453524/what-is-causing-this-error-com-android-tools-r8-internal-nc-sealed-classes-are

Does this mean that we need to wait for R8 to catch up and support the proper sealed classes of kotlin?
Can Keeper use a different R8 than the one bundled with AGP?

**Comment by ZacSweers** [comment, marked as answer]
Not sure R8 supports that yet. We use JDK 19 but still target java 11.

**Reply by igor-budi** [reply]
Yes, I see it here
https://github.com/slackhq/keeper/commit/5b9d59cccf2d9b38e204f8919fdb4be6c01ef5c8

Did the same on my project and it seems to be working well.

**Reply by igor-budi** [reply]
@ZacSweers Google fixed this but it is not in the main AGP yet

https://issuetracker.google.com/issues/227160052#comment37

applied the new R8 to the project but still see the issue with keeper 0.15

**Reply by ZacSweers** [reply]
Make sure you set it to the `keeperR8` configuration. We did that and are targeting Java 17 now.
